### PR TITLE
[Feature] Markdown Support

### LINKED ISSUE #26: Markdown Support
Instead of rewriting the message, I would like to write the text version in markdown, and have the html version auto-generated. Taking care of the multipart boundaries by hand and rewriting emails in HTML is just 😿

### PR DESCRIPTION
**Pull request by seshrs**
This PR adds Markdown support to `mailmerge` through a new command-line option `--markdown`. Text from non-multipart email templates will be converted to Markdown using [Python-Markdown](https://github.com/Python-Markdown/markdown).

- Added documentation (open to suggestions to make it clearer).
- Added tests. Verified that all tests pass locally.
- Manually sent an email using markdown and attachments and verified that they appear correctly on Gmail.
- Fixed a small bug in the attachments feature. Refactored some shared code in both `api.py` and the tests.

Closes #26.

### CODE REVIEW

File: mailmerge/api.py [hunk 1]
```diff
@@ -94,6 +91,39 @@ def addattachments(message, template_path):
...
         message = multipart_message
+    # HACK: For Python2 (see comments in `_create_boundary`)
+    message = _create_boundary(message)
+    return message
+
+
+def convert_markdown(message):
+    """Convert markdown in message text to HTML."""
+    if message.is_multipart():
+        print(">>> Warning: Markdown conversion requires non-multipart email.")
```
Comment (awdeorio) [inline mailmerge/api.py]: Should this exit?

File: mailmerge/api.py [hunk 2]
```diff
@@ -121,7 +151,7 @@ def addattachments(message, template_path):
         message.attach(part)
         print(">>> attached {}".format(normalized_path))
 
-    del message['attachments']
+    del message['attachment']
```
Comment (awdeorio) [inline mailmerge/api.py]: Is this a bugfix?

File: tests/test_markdown.py [hunk 3]
```diff
@@ -0,0 +1,51 @@
+"""Test messages with markdown."""
```
Comment (awdeorio) [inline tests/test_markdown.py]: thanks for adding a test!

File: tests/test_smtp_base.py [hunk 4]
```diff
@@ -0,0 +1,19 @@
+"""Base test class for SMTP tests."""
```
Comment (awdeorio) [inline tests/test_smtp_base.py]: Good

File: mailmerge/api.py (hunk 2 above)
Comment (seshrs) [reply mailmerge/api.py]: Yes, it is. It probably should go in a separate PR. Sorry about that! Will make that change.

**Comment by seshrs** [comment]
In addition to the user specifying markdown configurations, maybe mailmerge can auto-detect Markdown in non-multipart emails? (Specifically, look for hyperlinks, underscores or asterisks in the message?) what do you think of that?

**Comment by awdeorio** [comment]
> maybe mailmerge can auto-detect Markdown in non-multipart emails? (Specifically, look for hyperlinks, underscores or asterisks in the message?) what do you think of that?

Hmm, a heuristic would be interesting.  However, I'm a little leery of guessing wrong.  For example, someone might use *emphasis* in their email, but not really intend to trigger Markdown formatting.

> Another option is to make Markdown conversion default (because even if there is no markdown, emails will still be rendered properly), and provide a --no-markdown option instead.

I believe that this would mean that every email sent by mailmerge would be in (dual) HTML format?  This could be surprising to users!

File: mailmerge/api.py (hunk 1 above)
Comment (seshrs) [reply mailmerge/api.py]: If we choose to use the custom header option, then this condition would never be met. Resolving in b99e752345d41e33a43558f121e29d6f69ab792e.

File: mailmerge/api.py (hunk 2 above)
Comment (seshrs) [reply mailmerge/api.py]: Removed in b99e752345d41e33a43558f121e29d6f69ab792e.

File: mailmerge/api.py [hunk 5]
```diff
@@ -94,6 +91,39 @@ def addattachments(message, template_path):
...
         multipart_message.attach(email.mime.text.MIMEText(original_text))
         message = multipart_message
+    # HACK: For Python2 (see comments in `_create_boundary`)
+    message = _create_boundary(message)
+    return message
+
+
+def convert_markdown(message):
+    """Convert markdown in message text to HTML."""
+    if not message['Content-Type'].startswith("text/markdown"):
```
Comment (awdeorio) [inline mailmerge/api.py]: I think the code would be more readable if the `if` statement was in the calling function.  Then, put an `assert` right here.

File: mailmerge/api.py [hunk 6]
```diff
@@ -94,6 +91,39 @@ def addattachments(message, template_path):
...
+        # the multipart message (as per RFC 2046)
+        if payload_item['Content-Type'].startswith('text/plain'):
+            original_text = payload_item.get_payload()
+            html_text = markdown.markdown(original_text)
+            html_payload = future.backports.email.mime.text.MIMEText(
+                "<html><body>{}</body></html>".format(html_text),
+                "html",
+            )
+            message.attach(html_payload)
+    return message
```
Comment (awdeorio) [inline mailmerge/api.py]: Should this function delete the custom header?

Comment (seshrs) [reply mailmerge/api.py]: Yes, it should. L103 deletes the header before converting the message to multipart.

File: mailmerge/api.py (hunk 5 above)
Comment (seshrs) [reply mailmerge/api.py]: Resolved in 3b7253653df27ce162198612c2c6d5f1ce2ccab4.

File: mailmerge/api.py (hunk 6 above)
Comment (awdeorio) [reply mailmerge/api.py]: Whoops, I missed that.

**Pull request merged**